akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala

```diff
@@ -359,127 +368,116 @@ object ShardCoordinator {
-        case _: ShardHomeDeallocated ⇒
-          persistentState = persistentState.updated(evt)
-      }
+  override def receive: Receive = waitingForState
 
-    case SnapshotOffer(_, state: State) ⇒
-      log.debug("receiveRecover SnapshotOffer {}", state)
-      //Old versions of the state object may not have unallocatedShard set,
-      // thus it will be null.
-      if (state.unallocatedShards == null)
-        persistentState = state.copy(unallocatedShards = Set.empty)
-      else
-        persistentState = state
-
-    case RecoveryCompleted ⇒
-      persistentState.regionProxies.foreach(context.watch)
-      persistentState.regions.foreach { case (a, _) ⇒ context.watch(a) }
-      persistentState.shards.foreach { case (a, r) ⇒ sendHostShardMsg(a, r) }
+  def waitingForState: Receive = {
+    case g @ GetSuccess(CoordinatorStateKey, _) ⇒
+      state = g.get(CoordinatorStateKey).value
+      state.regionProxies.foreach(context.watch)
+      state.regions.foreach { case (a, _) ⇒ context.watch(a) }
+      state.shards.foreach { case (a, r) ⇒ sendHostShardMsg(a, r) }
       allocateShardHomes()
+      activate()
+
+    case GetFailure(CoordinatorStateKey, _) ⇒
+      log.error(
+        "The ShardCoordinator was unable to get an initial state within 'waiting-for-state-timeout' (was retried): {} millis",
+        waitingForStateTimeout.toMillis)
+      getState()
+
+    case NotFound(CoordinatorStateKey, _) ⇒ activate()
+    case m                                ⇒ stash()
   }
 
-  override def receiveCommand: Receive = {
+  def active: Receive = {
+    case UpdateSuccess(CoordinatorStateKey, Some(evt: DomainEvent)) ⇒
+      log.debug("The coordinator state was successfully updated to {}", evt)
+
+    case UpdateTimeout(CoordinatorStateKey, Some(evt: DomainEvent)) ⇒
+      log.error(
+        "The ShardCoordinator was unable to update a distributed state within 'updating-state-timeout' (was retried): {} millis",
+        updatingStateTimeout.toMillis)
+      sendUpdate(evt)
+
+    case ModifyFailure(CoordinatorStateKey, error, cause, Some(evt: DomainEvent)) ⇒
+      log.error(
+        cause,
+        "The ShardCoordinator was unable to update a distributed state with error {}. Was retried with event {}.",
+        error,
+        evt)
+      sendUpdate(evt)
+
     case Register(region) ⇒
       log.debug("ShardRegion registered: [{}]", region)
       aliveRegions += region
-      if (persistentState.regions.contains(region))
+      if (state.regions.contains(region))
         sender() ! RegisterAck(self)
       else {
         gracefulShutdownInProgress -= region
-        saveSnapshotWhenNeeded()
-        persist(ShardRegionRegistered(region)) { evt ⇒
-          val firstRegion = persistentState.regions.isEmpty
+        val firstRegion = state.regions.isEmpty
 
-          persistentState = persistentState.updated(evt)
-          context.watch(region)
-          sender() ! RegisterAck(self)
+        update(ShardRegionRegistered(region))
```

Comment:
Perhaps continue with watch and RegisterAck when you receive the `UpdateSuccess`? You can use the request (context) parameter in the `Update` message to pass along e.g. `sender`.

If you receive `UpdateTimeout` we should perhaps not retry the update here, since the region also has a retry mechanism.